crates/worker/src/lib.rs

```diff
@@ -331,68 +317,241 @@ impl Worker {
+            metadata.my_node_id(),
+            self.options,
+            self.rocksdb_storage,
+            self.networking,
+            bifrost,
+            invoker_handle,
+        );
+
+        loop {
+            tokio::select! {
+                _ = &mut shutdown => {
+                    debug!("Initiating shutdown of worker");
+
+                    // This will only shutdown rocksdb writer thread. Everything else will respond to
+                    // the cancellation signal independently.
+                    shutdown_signal.drain().await;
+                    break;
+                },
+                epoch_table_update = metadata.wait_for_version(MetadataKind::EpochTable, latest_epoch_table_version) => {
+                    latest_epoch_table_version = epoch_table_update?.next();
+                    partition_processor_manager.on_epoch_table_update().await?
                 },
-            )?;
+                partition_table_update = metadata.wait_for_version(MetadataKind::PartitionTable, latest_partition_table_version) => {
+                    latest_partition_table_version = partition_table_update?.next();
+                    partition_processor_manager.on_partition_table_update().await?
+                },
+            }
         }
 
-        // Invoker service
-        tc.spawn_child(TaskKind::SystemService, "invoker", None, self.invoker.run())?;
+        Ok(())
+    }
+}
+
+struct PartitionProcessorManager {
+    node_id: GenerationalNodeId,
+    running_partition_processors: HashMap<PartitionId, (TaskId, RangeInclusive<PartitionKey>)>,
+    last_applied_partition_table_version: Version,
+    last_applied_epoch_table_version: Version,
 
-        tokio::select! {
-            _ = shutdown => {
-                debug!("Initiating shutdown of worker");
+    options: Options,
+    rocksdb_storage: RocksDBStorage,
+    networking: Networking,
+    bifrost: Bifrost,
+    invoker_handle: ChannelServiceHandle,
+}
 
-                // This will only shutdown rocksdb writer thread. Everything else will respond to
-                // the cancellation signal independently.
-                shutdown_signal.drain().await;
+impl PartitionProcessorManager {
+    fn new(
+        node_id: GenerationalNodeId,
+        options: Options,
+        rocksdb_storage: RocksDBStorage,
+        networking: Networking,
+        bifrost: Bifrost,
+        invoker_handle: ChannelServiceHandle,
+    ) -> Self {
+        Self {
+            running_partition_processors: HashMap::default(),
+            last_applied_epoch_table_version: Version::INVALID,
+            last_applied_partition_table_version: Version::INVALID,
+            node_id,
+            options,
+            rocksdb_storage,
+            networking,
+            bifrost,
+            invoker_handle,
+        }
+    }
+
+    #[allow(clippy::map_entry)]
+    async fn on_partition_table_update(&mut self) -> anyhow::Result<()> {
```

Comment:
At the moment, this assumes that we should start/stop partitions based on information in the metadata, is this because we don't have attachment plan yet, or is this how you think we should do that long term as well?

It's more of a question on whether this component should act declaratively (on metadata observations) or imperatively (based on controller plans)

Reply:
After an offline (actually online) discussion we concluded that the cluster controller will send imperative commands to change the state of a worker instead of communicating control decisions via the metadata. One strong benefit of this approach is that we don't have to write to the metadata store for every control decision.